sdk/core/azure-core/src/main/java/com/azure/core/http/rest/SyncRestProxy.java

```diff
@@ -0,0 +1,692 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.http.rest;
+
+import com.azure.core.exception.HttpResponseException;
+import com.azure.core.exception.UnexpectedLengthException;
+import com.azure.core.http.*;
+import com.azure.core.http.policy.CookiePolicy;
+import com.azure.core.http.policy.HttpPipelinePolicy;
+import com.azure.core.http.policy.RetryPolicy;
+import com.azure.core.http.policy.UserAgentPolicy;
+import com.azure.core.implementation.AccessibleByteArrayOutputStream;
+import com.azure.core.implementation.TypeUtil;
+import com.azure.core.implementation.http.UnexpectedExceptionInformation;
+import com.azure.core.implementation.serializer.HttpResponseDecoder;
+import com.azure.core.implementation.serializer.HttpResponseDecoder.HttpDecodedResponse;
+import com.azure.core.util.*;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.core.util.serializer.JacksonAdapter;
+import com.azure.core.util.serializer.SerializerAdapter;
+import com.azure.core.util.serializer.SerializerEncoding;
+import com.azure.core.util.tracing.Tracer;
+import com.azure.core.util.tracing.TracerProxy;
+import reactor.core.Exceptions;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Signal;
+import reactor.util.context.ContextView;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.invoke.MethodHandle;
+import java.lang.reflect.*;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static com.azure.core.implementation.serializer.HttpResponseBodyDecoder.shouldEagerlyReadResponse;
+
+/**
+ * Type to create a proxy implementation for an interface describing REST API methods.
+ *
+ * RestProxy can create proxy implementations for interfaces with methods that return deserialized Java objects as well
+ * as asynchronous Single objects that resolve to a deserialized Java object.
+ */
+public final class SyncRestProxy implements InvocationHandler {
```

Comment:
If we do end up with a sync class for `RestProxy` I wonder if we should add an abstract class on top of RestProxy and SyncRestProxy to ensure common functionality is maintained. This will also let us play around with mocking for unit tests and we can have package-private methods for common functionality such as creating the HttpRequest, starting and ending tracing, and other shared functionality.